movici_simulation_core/models/rules/expression.py

```diff
@@ -0,0 +1,347 @@
+        elif self.operator == "OR":
+            return any(op.evaluate(simtime, clocktime, attributes) for op in self.operands)
+        return False
+
+    def get_attribute_names(self) -> set[str]:
+        """Return set of all attribute names referenced in this expression.
+
+        :returns: Set of attribute names
+        :rtype: set[str]
+        """
+        names: set[str] = set()
+        for operand in self.operands:
+            names.update(operand.get_attribute_names())
+        return names
+
+
+ParsedCondition = Union[Comparison, BooleanExpression]
+
+
+class _NotExpr:
+    """Helper class for NOT expression parsing."""
+
+    def __init__(self, tokens: pp.ParseResults) -> None:
+        self.operand = tokens[0][1]
+
+    def to_expr(self) -> BooleanExpression:
+        """Convert to BooleanExpression.
+
+        :returns: BooleanExpression with NOT operator
+        :rtype: BooleanExpression
+        """
+        operand = self.operand
+        if hasattr(operand, "to_expr"):
+            operand = operand.to_expr()
+        return BooleanExpression(operator="NOT", operands=[operand])
+
+
+class _BinaryExpr:
+    """Helper class for binary expression parsing."""
+
+    def __init__(self, tokens: pp.ParseResults) -> None:
+        self.tokens = tokens[0]
+
+    def to_expr(self) -> Union[Comparison, BooleanExpression]:
+        """Convert to BooleanExpression or Comparison.
+
+        :returns: Parsed expression
+        :rtype: Union[Comparison, BooleanExpression]
+        """
+        operands = []
+        operator = None
+
+        for i, token in enumerate(self.tokens):
+            if i % 2 == 0:
+                if hasattr(token, "to_expr"):
+                    operands.append(token.to_expr())
+                else:
+                    operands.append(token)
+            else:
+                op = token.upper() if isinstance(token, str) else token
+                if op in ("&", "AND"):
+                    operator = "AND"
+                elif op in ("|", "OR"):
+                    operator = "OR"
+
+        if len(operands) == 1:
+            return operands[0]
+
+        return BooleanExpression(operator=operator, operands=operands)
+
+
+def _build_grammar() -> pp.ParserElement:
+    """Build the pyparsing grammar for condition expressions.
+
+    :returns: Pyparsing grammar for parsing condition expressions
+    :rtype: pp.ParserElement
+    """
+    integer = pp.Word(pp.nums).setParseAction(lambda t: int(t[0]))
+    decimal = pp.Combine(pp.Optional(pp.Word(pp.nums)) + "." + pp.Word(pp.nums)).setParseAction(
+        lambda t: float(t[0])
+    )
+    number = decimal | integer
+
+    # Time expressions: durations (34h, 5m, 1d5h30s) and clock times (12:30)
+    time_unit = pp.oneOf("h m s d")
+    single_duration = pp.Combine(pp.Word(pp.nums) + time_unit)
+    duration = pp.Combine(pp.OneOrMore(single_duration)).setParseAction(
+        lambda t: parse_time_value(t[0])
+    )
```

Comment:
If you rewrite this, so that you get a list of tuples in a duration node (eg: `[(1, "h"),(2, "m")]`), it becomes straight forward to calculate the duration without the `timelength` package.